### PR TITLE
`EpochIterator`, `TensorFlowTrainer` : refactoring and fixes

### PR DESCRIPTION
**Pull request by nicolaspi**
Fixes https://github.com/keras-team/keras/issues/20394
Fixes https://github.com/keras-team/keras/issues/20390
Fixes https://github.com/keras-team/keras/issues/20344

`TensorFlowTrainer` :

   - Added usage of `iterator.get_next_as_optional()`, which enables `steps_per_execution` to exceed the previous limit of 32 while preventing `tf.errors.OutOfRangeError`

`EpochIterator` :

   - Refactored to unify code base across different backends
   - Fixed edge cases involving various combinations of `steps_per_epoch` and `steps_per_execution`

### CODE REVIEW

File: keras/src/trainers/epoch_iterator.py [hunk 1]
```diff
@@ -75,55 +75,86 @@ def __init__(
...
     @property
     def num_batches(self):
         if self.steps_per_epoch:
             return self.steps_per_epoch
         # Either copied from the data_adapter, or
         # inferred at the end of an iteration.
         return self._num_batches
+
+
+class EpochIterator(_EpochIterator):
```
Comment (fchollet) [inline keras/src/trainers/epoch_iterator.py:149]: Couldn't the two classes be merged into one?

File: keras/src/trainers/epoch_iterator.py [hunk 2]
```diff
@@ -75,55 +75,86 @@ def __init__(
...
+    def __next__(self):
+        buffer = []
+        step, iterator = super().__next__()
```
Comment (fchollet) [inline keras/src/trainers/epoch_iterator.py:152]: Just need to call `next(self._epoch_iterator)`

File: keras/src/trainers/epoch_iterator.py [hunk 3]
```diff
@@ -75,55 +75,86 @@ def __init__(
     def _get_iterator(self):
         return self.data_adapter.get_numpy_iterator()
 
-    def enumerate_epoch(self):
```
Comment (fchollet) [inline keras/src/trainers/epoch_iterator.py:78]: This method must be kept for backwards compatibility (for users who have a custom fit())

**Review by fchollet — COMMENTED** [review]
This is neat -- thanks for the PR.

**Comment by fchollet** [comment]
It's fairly confusing to have backend-specific iterators subclass `_EpochIterator` -- much better to have a single `EpochIterator` class and have backend-specific classes override what they need to change.

File: keras/src/trainers/epoch_iterator.py (hunk 1 above)
Comment (nicolaspi) [reply keras/src/trainers/epoch_iterator.py]: Yes, I merged them.

File: keras/src/trainers/epoch_iterator.py (hunk 2 above)
Comment (nicolaspi) [reply keras/src/trainers/epoch_iterator.py]: Indeed

File: keras/src/trainers/epoch_iterator.py (hunk 3 above)
Comment (nicolaspi) [reply keras/src/trainers/epoch_iterator.py:78]: It's back.

**Comment by nicolaspi** [comment]
> It's fairly confusing to have backend-specific iterators subclass `_EpochIterator` -- much better to have a single `EpochIterator` class and have backend-specific classes override what they need to change.

I removed `_EpochIterator`.

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution!

**Pull request merged**